architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,677 @@
+   - Implement backup storage mechanism in Kubernetes (ConfigMaps or PVs depending on size).
+   - This task depends on item 4 (interfaces) and blocks item 6 (CLI implementation).
+
+6. **Upgrade Lock Mechanism**
+
+   - Implement the upgrade lock interface to prevent concurrent modifications.
+   - Update existing CLI commands to check for locks before data modification.
+   - Can be implemented in parallel with items 1-5. Required for item 7.
+
+7. **CLI Command Implementation**
+
+   - Implement the `rad upgrade kubernetes` command, integrating all previously defined components and interfaces.
+   - Ensure the command performs pre-flight checks, user data backup creation, component upgrades, rollback on failure, and post-upgrade verification.
+   - Include detailed CLI output and logging for user visibility.
+   - Add necessary unit and functional tests to validate command behavior.
+   - This task depends on all previous tasks (1-6) and should be implemented last.
+
+### Out of Scope for Implementation
+
+- **Dry-run functionality**: After team discussion, the dry-run feature was explicitly excluded from this implementation.
+- **Dependency upgrades**: Upgrading major versions of dependencies (Contour, Dapr, Postgres) is explicitly out of scope and should be handled separately.
+- **Automatic CLI upgrades**: Updating the local Radius CLI automatically is not included; users must manually update their CLI version separately.
+
+### Implementation Risks and Mitigations
+
+- **Backup Reliability**: User data backup and restore mechanisms must be thoroughly tested to ensure reliability. Consider edge cases such as backup corruption or restoration failures.
+- **Lock Persistence**: Ensure upgrade locks have proper timeout mechanisms to avoid permanently locked systems if a process terminates unexpectedly.
+
+### Testing Strategy
+
+- **Unit Tests**: Cover all new code paths, especially backup and restore logic, upgrade logic, and error handling.
+- **Functional Tests**: Validate end-to-end upgrade scenarios, including successful upgrades, upgrades with custom configurations, failure scenarios, and rollback procedures.
+- **Compatibility Tests**: Verify compatibility between different Radius CLI versions and control plane components.
+
+## Open Questions
+
+- **Rollback reliability**: What specific scenarios could cause rollback to fail, and how should we mitigate these risks?
+  - Do we need rollbacks in version 1?
+  - If the answer to the item above is yes, then do we need to add the implementation of the interfaces for Postgres in version 1?
+- **Version skipping limits**: Should we enforce incremental upgrades for certain major version jumps, or always allow direct version skipping?
+- **Upgrade notifications**: How should we notify users clearly about the CLI version mismatch after upgrading the control plane?
+- **Resource constraints**: How do we handle scenarios where the cluster lacks sufficient resources to perform a rolling upgrade?
+
```

Comment:
One thing that I'm not seeing mentioned is how we handle breaking changes. we're still in 0.x versions, and sometimes breaking changes might happen in the resource specs or in other places